src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,130 @@
+/* eslint-disable jsx-a11y/control-has-associated-label */
+import {
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+import classNames from 'classnames';
+import { Todo, ReducerType } from '../../types';
+import { DispatchContext } from '../../state/TodosContext';
+
+interface Props {
+  todoItem: Todo,
+}
+
+export const TodoItem: React.FC<Props> = ({ todoItem }) => {
+  const [edit, setEdit] = useState(false);
+  const [todo, setTodo] = useState(todoItem);
+  const [initialTitle, setInitialTitle] = useState(todo.title);
+
+  const inputRef = useRef<HTMLInputElement | null>(null);
+  const dispatch = useContext(DispatchContext);
+
+  useEffect(() => {
+    setTodo(todoItem);
+  }, [todoItem]);
+
+  const { id, title, completed } = todo;
+
+  const handleInputSubmit = () => {
+    setEdit(false);
+
+    if (title.trim()) {
+      dispatch({
+        type: ReducerType.ChangeTodo,
+        payload: {
+          id,
+          title: title.trim(),
+          completed,
+        },
+      });
+    } else {
+      dispatch({
+        type: ReducerType.DeleteTodo,
+        payload: id,
+      });
+    }
+  };
+
+  const handleKeyDown = (
+    event: React.KeyboardEvent<HTMLInputElement>,
+  ) => {
+    if (event.key === 'Enter') {
+      handleInputSubmit();
+    }
+
+    if (event.key === 'Escape') {
+      setEdit(false);
+      setTodo({ ...todo, title: initialTitle });
+    }
+  };
+
+  useEffect(() => {
+    setInitialTitle(todo.title);
+  }, [todoItem]);
+
+  const toggleCompleted = () => {
+    dispatch({
+      type: ReducerType.ChangeTodo,
+      payload: {
+        id,
+        title,
+        completed: !todo.completed,
+      },
+    });
+  };
+
+  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTodo(currentTodo => ({
```

Comment:
You are creating a new object every time the input changes. This can affect performance, especially with a large number of items. Consider using a simpler state update.